File: haskell/src/Surd/Trig/RootOfUnity.hs
-- | Express roots of unity as radical expressions.
--
-- A primitive nth root of unity ζₙ = e^(2πi/n) can ALWAYS be expressed
-- in radicals, since cyclotomic extensions have abelian Galois groups
-- (hence solvable by the Kronecker-Weber theorem).
--
-- The compass-and-straightedge constructible case (n = 2^a · distinct
-- Fermat primes) is the special case where only square roots are needed.
-- For general n, the radical expression may involve roots of higher
-- degree (cube roots for factors of 3 in φ(n), fifth roots for factors
-- of 5, etc.) and may pass through complex intermediates due to the
-- casus irreducibilis.
module Surd.Trig.RootOfUnity
  ( cosOfUnity
  , sinOfUnity
  , isConstructible
  , fermatPrimes
  ) where

import Data.Bits qualified as Bits
import Data.Ratio ((%))
import Surd.Types
import Math.Internal.Positive (Positive)
import Math.Internal.PrimeFactors (factorise, isPrime)
import Surd.Trig.Galois (cosOfUnityViaGauss)

-- | Known Fermat primes (these are the only known ones, and it's
-- conjectured there are no others).
fermatPrimes :: [Integer]
fermatPrimes = [3, 5, 17, 257, 65537]

-- | Check if cos(2π/n) is expressible using only nested square roots
-- (i.e., compass-and-straightedge constructible).
--
-- This is true iff n = 2^a · product of distinct Fermat primes.
--
-- Note: cos(2π/n) is expressible in radicals (possibly involving
-- higher-degree roots) for ALL n, not just constructible ones.
isConstructible :: Int -> Bool
isConstructible n =
  let fs = factorise (fromIntegral n :: Positive)
      oddFactors = [(p, e) | (p, e) <- fs, p /= 2]
  in all (\(p, e) -> e == 1 && p `elem` fermatPrimes) oddFactors

-- | Compute cos(2π/n) as a radical expression.
--
-- Always succeeds for n ≥ 1. For small n and constructible cases,
-- returns optimised closed-form expressions. For general n, uses
-- the Gauss period descent which may involve complex intermediates.
cosOfUnity :: Int -> Maybe (RadExpr Rational)
cosOfUnity n
  | n <= 0    = Nothing
  -- Small exact cases (hand-optimised)
  | n == 1    = Just (Lit 1)
  | n == 2    = Just (Lit (-1))
  | n == 3    = Just (Lit (-1 % 2))
  | n == 4    = Just (Lit 0)
  | n == 5    = Just cos2piOver5
  | n == 6    = Just (Lit (1 % 2))
  | n == 8    = Just (Mul (Inv (Lit 2)) (Root 2 (Lit 2)))
  | n == 10   = Just cos2piOver10
  | n == 12   = Just (Mul (Inv (Lit 2)) (Root 2 (Lit 3)))
  | n == 15   = Just cos2piOver15
  | n == 16   = Just cos2piOver16
  | n == 17   = Just cos2piOver17
  | n == 20   = Just cos2piOver20
  | n == 24   = Just cos2piOver24
  -- Powers of 2: half-angle recurrence (only square roots)
  | isPowerOf2 n = Just (cosOfPow2 n)
  -- Odd prime or odd prime power: Gauss period descent
  -- (Z/p^k Z)* is cyclic for odd primes, so this always works
  | isPrime (fromIntegral n) = cosOfUnityViaGauss n
  | Just (p, k) <- primePowerDecomp n, p > 2, k > 1 =
      cosOfUnityViaGauss n
  -- General composite: decompose via CRT / product formulas
  | otherwise = cosOfUnityComposite n

-- | Compute sin(2π/n) as a radical expression.
sinOfUnity :: Int -> Maybe (RadExpr Rational)
sinOfUnity n
  | n <= 0    = Nothing
  | n == 1    = Just (Lit 0)
  | n == 2    = Just (Lit 0)
  | n == 3    = Just (Mul (Lit (1 % 2)) (Root 2 (Lit 3)))
  | n == 4    = Just (Lit 1)
  | n == 6    = Just (Mul (Lit (1 % 2)) (Root 2 (Lit 3)))
  | n == 8    = Just (Mul (Inv (Lit 2)) (Root 2 (Lit 2)))
  | n == 12   = Just (Lit (1 % 2))
  | otherwise =
      -- sin(2π/n) = √(1 - cos²(2π/n))
      case cosOfUnity n of
        Nothing -> Nothing
        Just c  ->
          let sin2 = Add (Lit 1) (Neg (Mul c c))
          in Just (Root 2 sin2)

-- cos(2π/5) = (√5 - 1) / 4
cos2piOver5 :: RadExpr Rational
cos2piOver5 = Mul (Inv (Lit 4)) (Add (Root 2 (Lit 5)) (Lit (-1)))

-- cos(2π/10) = cos(π/5) = (1 + √5) / 4
cos2piOver10 :: RadExpr Rational
cos2piOver10 = Mul (Inv (Lit 4)) (Add (Lit 1) (Root 2 (Lit 5)))

-- cos(2π/15) = (1 + √5 + √(30 - 6√5)) / 8
cos2piOver15 :: RadExpr Rational
cos2piOver15 =
  Mul (Inv (Lit 8))
    (Add (Add (Lit 1) (Root 2 (Lit 5)))
         (Root 2 (Add (Lit 30) (Neg (Mul (Lit 6) (Root 2 (Lit 5)))))))

-- cos(2π/16) = cos(π/8) = √(2 + √2) / 2
cos2piOver16 :: RadExpr Rational
cos2piOver16 =
  Mul (Inv (Lit 2)) (Root 2 (Add (Lit 2) (Root 2 (Lit 2))))

-- cos(2π/17): the famous 17-gon (Gauss)
-- 16·cos(2π/17) = -1 + √17 + √(34-2√17) + 2·√(17+3√17-√(34-2√17)-2·√(34+2√17))
cos2piOver17 :: RadExpr Rational
cos2piOver17 =
  let s17 = Root 2 (Lit 17)
      a = Root 2 (Add (Lit 34) (Neg (Mul (Lit 2) s17)))
      b = Root 2 (Add (Lit 34) (Mul (Lit 2) s17))
  in Mul (Inv (Lit 16))
       (Add (Add (Add (Lit (-1)) s17) a)
            (Mul (Lit 2) (Root 2
              (Add (Add (Lit 17) (Mul (Lit 3) s17))
                   (Add (Neg a) (Neg (Mul (Lit 2) b)))))))

-- cos(2π/20) = cos(π/10) = √(10 + 2√5) / 4
cos2piOver20 :: RadExpr Rational
cos2piOver20 =
  Mul (Inv (Lit 4)) (Root 2 (Add (Lit 10) (Mul (Lit 2) (Root 2 (Lit 5)))))

-- cos(2π/24) = cos(π/12) = (√6 + √2) / 4
cos2piOver24 :: RadExpr Rational
cos2piOver24 =
  Mul (Inv (Lit 4)) (Add (Root 2 (Lit 6)) (Root 2 (Lit 2)))

-- | cos(2π/2^k) via the half-angle recurrence:
-- cos(2π/2^(k+1)) = √((1 + cos(2π/2^k)) / 2)
cosOfPow2 :: Int -> RadExpr Rational
cosOfPow2 n = go n
  where
    go 1 = Lit 1
    go 2 = Lit (-1)
    go 4 = Lit 0
    go k =
      let half = go (k `div` 2)
      in Root 2 (Mul (Inv (Lit 2)) (Add (Lit 1) half))

-- | Decompose n into (p, k) if n = p^k for prime p, k ≥ 1.
primePowerDecomp :: Int -> Maybe (Int, Int)
primePowerDecomp n =
  let fs = factorise (fromIntegral n :: Positive)
  in case fs of
    [(p, k)] -> Just (fromIntegral p, k)
    _        -> Nothing

-- | cos(2π/p^k) for prime p and k ≥ 2.
--
-- Uses the identity: cos(2π/p^k) is a root of the p-th Chebyshev-like
-- polynomial applied to cos(2π/p^{k-1}), specifically:
-- The minimal polynomial of 2cos(2π/p^k) over Q(2cos(2π/p^{k-1}))
-- has degree p and can be solved in radicals (the Galois group is cyclic).
--
-- For p = 2, this reduces to the half-angle formula (square root).
-- For odd p, this involves p-th roots.
-- | cos(2π/n) for composite n.
--
-- Strategy: factor n = n₁·n₂ where gcd(n₁, n₂) = 1, and use
-- the identity cos(2π/n) = cos(2π/(n₁·n₂)).
-- By CRT, there exist a, b with a·n₂ + b·n₁ = 1, so
-- 2π/n = a·(2π/n₁) + b·(2π/n₂), and we can use the
-- addition formula for cosine.
cosOfUnityComposite :: Int -> Maybe (RadExpr Rational)
cosOfUnityComposite n =
  let fs = factorise (fromIntegral n :: Positive)
  in case fs of
    []  -> Nothing
    [_] -> Nothing  -- prime power, handled elsewhere
    _   ->
      -- Split into two coprime factors n = n₁·n₂ with gcd(n₁,n₂) = 1.
      -- By CRT: a·n₁ + b·n₂ = 1 for integers a, b.
      -- Then 1/(n₁·n₂) = a/n₂ + b/n₁, so:
      --   2π/n = 2πa/n₂ + 2πb/n₁
      --   cos(2π/n) = cos(2πa/n₂ + 2πb/n₁)
      --             = cos(2πa/n₂)·cos(2πb/n₁) - sin(2πa/n₂)·sin(2πb/n₁)
      --
      -- Chebyshev T_k(cos θ) = cos(kθ) gives cos(2πk/m) from cos(2π/m).
      -- sinFromCos gives |sin|; the sign is determined from the quadrant.
      let ((p1, e1):_) = fs
          n1 = fromIntegral p1 ^ e1
          n2 = n `div` n1
          (a, b) = extGcdInt n1 n2  -- a*n1 + b*n2 = 1
          -- Reduce to canonical range
          a' = a `mod` n2
          b' = b `mod` n1
          -- Determine signs of sin(2πa'/n₂) and sin(2πb'/n₁)
          sA = sinSignFromAngle a' n2
          sB = sinSignFromAngle b' n1
          signFactor = sA * sB
      in do
        cosBase1 <- cosOfUnity n2  -- cos(2π/n₂)
        cosBase2 <- cosOfUnity n1  -- cos(2π/n₁)
        let cosAE = chebyshevSimple a' cosBase1  -- cos(2πa'/n₂)
            cosBE = chebyshevSimple b' cosBase2  -- cos(2πb'/n₁)
        if signFactor == 0
          then -- One of the sines is zero; cos(α+β) = cos(α)·cos(β)
            Just (Mul cosAE cosBE)
          else do
            let sinAbsA = sinFromCos cosAE  -- |sin(2πa'/n₂)|
                sinAbsB = sinFromCos cosBE  -- |sin(2πb'/n₁)|
                -- cos(α+β) = cos(α)cos(β) - signFactor · |sin(α)|·|sin(β)|
                sinProduct = Mul sinAbsA sinAbsB
                result = if signFactor == 1
                         then Add (Mul cosAE cosBE) (Neg sinProduct)
                         else Add (Mul cosAE cosBE) sinProduct
            Just result

-- | Sign of sin(2πk/m) for k ∈ [0, m-1]:
--   0 < k/m < 1/2 → positive (+1)
--   k/m = 0 or 1/2 → zero (0)
--   1/2 < k/m < 1 → negative (-1)
sinSignFromAngle :: Int -> Int -> Int
sinSignFromAngle k m
  | k' == 0       = 0
  | 2 * k' < m    = 1
  | 2 * k' == m   = 0
  | otherwise      = -1
  where k' = k `mod` m

-- | Extended GCD for integers: returns (a, b) such that a*x + b*y = gcd(x,y).
extGcdInt :: Int -> Int -> (Int, Int)
extGcdInt 0 _ = (0, 1)
extGcdInt x y =
  let (q, r) = y `divMod` x
      (a, b) = extGcdInt r x
  in (b - q * a, a)

-- | Simple Chebyshev: T_k(cos θ) = cos(kθ).
-- Computes T_k(x) symbolically using the recurrence.
chebyshevSimple :: Int -> RadExpr Rational -> RadExpr Rational
chebyshevSimple 0 _ = Lit 1
chebyshevSimple 1 x = x
chebyshevSimple k x = go 2 (Lit 1) x
  where
    go n t0 t1
      | n > k     = t1
      | otherwise  =
          let t2 = Add (Mul (Mul (Lit 2) x) t1) (Neg t0)
          in go (n + 1) t1 t2

-- | sin from cos: sin(θ) = √(1 - cos²(θ)) (positive in first quadrant).
sinFromCos :: RadExpr Rational -> RadExpr Rational
sinFromCos c = Root 2 (Add (Lit 1) (Neg (Mul c c)))

isPowerOf2 :: Int -> Bool
isPowerOf2 n = n > 0 && (n Bits..&. (n - 1)) == 0

File: mercury/root_of_unity.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: root_of_unity.m
% Main author: nyc
%
% Express roots of unity as radical expressions.
%
% A primitive nth root of unity ζₙ = e^(2πi/n) can ALWAYS be expressed
% in radicals, since cyclotomic extensions have abelian Galois groups.
%
% The compass-and-straightedge constructible case (n = 2^a · distinct
% Fermat primes) is the special case where only square roots are needed.
%
%---------------------------------------------------------------------------%

:- module root_of_unity.
:- interface.

:- import_module integer.
:- import_module list.
:- import_module map.
:- import_module maybe.
:- import_module rad_expr.
:- import_module rational.

%---------------------------------------------------------------------------%

    % Compute cos(2π/n) as a radical expression.
    %
:- func cos_of_unity(int) = maybe(rad_expr(rational)).

    % Compute sin(2π/n) as a radical expression.
    %
:- func sin_of_unity(int) = maybe(rad_expr(rational)).

    % All cos(2πk/n) for k coprime to n, from Gauss periods.
    %
:- func all_cos_of_unity(int) = maybe(map(int, rad_expr(rational))).

    % All sin(2πk/n) for k coprime to n, from Gauss periods.
    %
:- func all_sin_of_unity(int) = maybe(map(int, rad_expr(rational))).

    % Check if cos(2π/n) is constructible (only square roots needed).
    %
:- pred is_constructible(int::in) is semidet.

    % Known Fermat primes.
    %
:- func fermat_primes = list(integer).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module positive.
:- import_module prime_factors.
:- import_module trig_galois.

%---------------------------------------------------------------------------%
% Fermat primes
%---------------------------------------------------------------------------%

fermat_primes = [integer(3), integer(5), integer(17),
                 integer(257), integer(65537)].

is_constructible(N) :-
    Fs = factorise(det_positive(integer(N))),
    list.all_true(
        ( pred(PP::in) is semidet :-
            ( if pp_prime(PP) = integer(2) then
                true
            else
                pp_exponent(PP) = 1,
                list.member(pp_prime(PP), fermat_primes)
            )
        ), Fs).

%---------------------------------------------------------------------------%
% cos(2π/n)
%---------------------------------------------------------------------------%

cos_of_unity(N) = Result :-
    ( if int.'=<'(N, 0) then
        Result = no
    else if N = 1 then
        Result = yes(re_lit(rational.one))
    else if N = 2 then
        Result = yes(re_lit(rational.'-'(rational.zero, rational.one)))
    else if N = 3 then
        Result = yes(re_lit(from_integers(integer.'-'(integer.zero,
            integer.one), integer(2))))
    else if N = 4 then
        Result = yes(re_lit(rational.zero))
    else if N = 5 then
        Result = yes(cos_2pi_over_5)
    else if N = 6 then
        Result = yes(re_lit(from_integers(integer.one, integer(2))))
    else if N = 8 then
        Result = yes(re_mul(re_inv(re_lit(rational(2))),
                            re_root(2, re_lit(rational(2)))))
    else if N = 10 then
        Result = yes(cos_2pi_over_10)
    else if N = 12 then
        Result = yes(re_mul(re_inv(re_lit(rational(2))),
                            re_root(2, re_lit(rational(3)))))
    else if N = 15 then
        Result = yes(cos_2pi_over_15)
    else if N = 16 then
        Result = yes(cos_2pi_over_16)
    else if N = 17 then
        Result = yes(cos_2pi_over_17)
    else if N = 20 then
        Result = yes(cos_2pi_over_20)
    else if N = 24 then
        Result = yes(cos_2pi_over_24)
    else if is_power_of_2(N) then
        Result = yes(cos_of_pow2(N))
    else if is_prime(integer(N)) then
        Result = cos_of_unity_via_gauss(N)
    else if prime_power_decomp(N) = yes({P, K}),
            int.'>'(P, 2), int.'>'(K, 1) then
        Result = cos_of_unity_via_gauss(N)
    else
        Result = cos_of_unity_composite(N)
    ).

%---------------------------------------------------------------------------%
% sin(2π/n)
%---------------------------------------------------------------------------%

sin_of_unity(N) = Result :-
    ( if int.'=<'(N, 0) then
        Result = no
    else if N = 1 then
        Result = yes(re_lit(rational.zero))
    else if N = 2 then
        Result = yes(re_lit(rational.zero))
    else if N = 3 then
        Result = yes(re_mul(re_lit(from_integers(integer.one, integer(2))),
                            re_root(2, re_lit(rational(3)))))
    else if N = 4 then
        Result = yes(re_lit(rational.one))
    else if N = 6 then
        Result = yes(re_mul(re_lit(from_integers(integer.one, integer(2))),
                            re_root(2, re_lit(rational(3)))))
    else if N = 8 then
        Result = yes(re_mul(re_inv(re_lit(rational(2))),
                            re_root(2, re_lit(rational(2)))))
    else if N = 12 then
        Result = yes(re_lit(from_integers(integer.one, integer(2))))
    else
        ( if cos_of_unity(N) = yes(C) then
            Sin2 = re_add(re_lit(rational.one),
                          re_neg(re_mul(C, C))),
            Result = yes(re_root(2, Sin2))
        else
            Result = no
        )
    ).

%---------------------------------------------------------------------------%
% Hardcoded small cases
%---------------------------------------------------------------------------%

:- func cos_2pi_over_5 = rad_expr(rational).
cos_2pi_over_5 = re_mul(re_inv(re_lit(rational(4))),
    re_add(re_root(2, re_lit(rational(5))),
           re_lit(rational(-1)))).

:- func cos_2pi_over_10 = rad_expr(rational).
cos_2pi_over_10 = re_mul(re_inv(re_lit(rational(4))),
    re_add(re_lit(rational.one), re_root(2, re_lit(rational(5))))).

:- func cos_2pi_over_15 = rad_expr(rational).
cos_2pi_over_15 =
    re_mul(re_inv(re_lit(rational(8))),
        re_add(
            re_add(re_lit(rational.one), re_root(2, re_lit(rational(5)))),
            re_root(2, re_add(re_lit(rational(30)),
                re_neg(re_mul(re_lit(rational(6)),
                              re_root(2, re_lit(rational(5))))))))).

:- func cos_2pi_over_16 = rad_expr(rational).
cos_2pi_over_16 =
    re_mul(re_inv(re_lit(rational(2))),
        re_root(2, re_add(re_lit(rational(2)),
                           re_root(2, re_lit(rational(2)))))).

:- func cos_2pi_over_17 = rad_expr(rational).
cos_2pi_over_17 =
    re_mul(re_inv(re_lit(rational(16))),
        re_add(
            re_add(
                re_add(re_lit(rational(-1)), S17),
                A17),
            re_mul(re_lit(rational(2)),
                re_root(2,
                    re_add(
                        re_add(re_lit(rational(17)),
                               re_mul(re_lit(rational(3)), S17)),
                        re_add(re_neg(A17),
                               re_neg(re_mul(re_lit(rational(2)), B17))))))
        )) :-
    S17 = re_root(2, re_lit(rational(17))),
    A17 = re_root(2, re_add(re_lit(rational(34)),
                              re_neg(re_mul(re_lit(rational(2)), S17)))),
    B17 = re_root(2, re_add(re_lit(rational(34)),
                              re_mul(re_lit(rational(2)), S17))).

:- func cos_2pi_over_20 = rad_expr(rational).
cos_2pi_over_20 =
    re_mul(re_inv(re_lit(rational(4))),
        re_root(2, re_add(re_lit(rational(10)),
            re_mul(re_lit(rational(2)),
                   re_root(2, re_lit(rational(5))))))).

:- func cos_2pi_over_24 = rad_expr(rational).
cos_2pi_over_24 =
    re_mul(re_inv(re_lit(rational(4))),
        re_add(re_root(2, re_lit(rational(6))),
               re_root(2, re_lit(rational(2))))).

%---------------------------------------------------------------------------%
% Powers of 2
%---------------------------------------------------------------------------%

:- pred is_power_of_2(int::in) is semidet.

is_power_of_2(N) :-
    int.'>'(N, 0),
    int.'/\\'(N, int.'-'(N, 1)) = 0.

:- func cos_of_pow2(int) = rad_expr(rational).

cos_of_pow2(K) = Result :-
    ( if K = 1 then
        Result = re_lit(rational.one)
    else if K = 2 then
        Result = re_lit(rational.'-'(rational.zero, rational.one))
    else if K = 4 then
        Result = re_lit(rational.zero)
    else
        Half = cos_of_pow2(int.'/'(K, 2)),
        Result = re_root(2, re_mul(re_inv(re_lit(rational(2))),
                                    re_add(re_lit(rational.one), Half)))
    ).

%---------------------------------------------------------------------------%
% Prime power decomposition
%---------------------------------------------------------------------------%

:- func prime_power_decomp(int) = maybe({int, int}).

prime_power_decomp(N) = Result :-
    Fs = factorise(det_positive(integer(N))),
    ( if Fs = [prime_power(P, K)] then
        Result = yes({integer.det_to_int(P), K})
    else
        Result = no
    ).

%---------------------------------------------------------------------------%
% Composite decomposition via CRT
%---------------------------------------------------------------------------%

:- func cos_of_unity_composite(int) = maybe(rad_expr(rational)).

cos_of_unity_composite(N) = Result :-
    Fs = factorise(det_positive(integer(N))),
    ( if Fs = [_ | _],
         int.'>='(list.length(Fs), 2) then
        Fs = [prime_power(P1, E1) | _],
        N1 = integer.det_to_int(integer_pow_int(P1, E1)),
        N2 = int.'/'(N, N1),
        ext_gcd_int(N1, N2, A0, B0),
        A = int.'mod'(A0, N2),
        B = int.'mod'(B0, N1),
        SA = sin_sign_from_angle(A, N2),
        SB = sin_sign_from_angle(B, N1),
        SignFactor = int.'*'(SA, SB),
        ( if cos_of_unity(N2) = yes(CosBase1),
             cos_of_unity(N1) = yes(CosBase2) then
            CosAE = chebyshev(A, CosBase1),
            CosBE = chebyshev(B, CosBase2),
            ( if SignFactor = 0 then
                Result = yes(re_mul(CosAE, CosBE))
            else
                SinAbsA = sin_from_cos(CosAE),
                SinAbsB = sin_from_cos(CosBE),
                SinProduct = re_mul(SinAbsA, SinAbsB),
                ( if SignFactor = 1 then
                    Result = yes(re_add(re_mul(CosAE, CosBE),
                                        re_neg(SinProduct)))
                else
                    Result = yes(re_add(re_mul(CosAE, CosBE),
                                        SinProduct))
                )
            )
        else
            Result = no
        )
    else
        Result = no
    ).

:- func integer_pow_int(integer, int) = integer.

integer_pow_int(B, E) = Result :-
    ( if int.'=<'(E, 0) then
        Result = integer.one
    else
        Result = integer.'*'(B, integer_pow_int(B, int.'-'(E, 1)))
    ).

:- func sin_sign_from_angle(int, int) = int.

sin_sign_from_angle(K, M) = Result :-
    K1 = int.'mod'(K, M),
    ( if K1 = 0 then
        Result = 0
    else if int.'<'(int.'*'(2, K1), M) then
        Result = 1
    else if int.'*'(2, K1) = M then
        Result = 0
    else
        Result = -1
    ).

:- pred ext_gcd_int(int::in, int::in, int::out, int::out) is det.

ext_gcd_int(X, Y, A, B) :-
    ( if X = 0 then
        A = 0,
        B = 1
    else
        Q = int.'/'(Y, X),
        R = int.'-'(Y, int.'*'(Q, X)),
        ext_gcd_int(R, X, A1, B1),
        A = int.'-'(B1, int.'*'(Q, A1)),
        B = A1
    ).

:- func sin_from_cos(rad_expr(rational)) = rad_expr(rational).

sin_from_cos(C) = re_root(2, re_add(re_lit(rational.one),
                                      re_neg(re_mul(C, C)))).

%---------------------------------------------------------------------------%
% All cos/sin of unity from Gauss periods
%---------------------------------------------------------------------------%

all_cos_of_unity(N) = Result :-
    ( if all_periods_via_gauss(N) = yes(Periods) then
        Result = yes(map.map_values(
            ( func(K, Pk) =
                re_mul(re_inv(re_lit(rational(2))),
                       re_add(Pk, Pnk)) :-
                NK = int.'-'(N, K),
                map.lookup(Periods, NK, Pnk)
            ), Periods))
    else
        Result = no
    ).

all_sin_of_unity(N) = Result :-
    ( if all_periods_via_gauss(N) = yes(Periods) then
        I = re_root(2, re_lit(rational.'-'(rational.zero, rational.one))),
        NegIOver2 = re_mul(re_inv(re_lit(rational(2))), re_neg(I)),
        Result = yes(map.map_values(
            ( func(K, Pk) =
                re_mul(NegIOver2, Diff) :-
                NK = int.'-'(N, K),
                map.lookup(Periods, NK, Pnk),
                Diff = re_add(Pk, re_neg(Pnk))
            ), Periods))
    else
        Result = no
    ).

%---------------------------------------------------------------------------%
:- end_module root_of_unity.
%---------------------------------------------------------------------------%
